src/interfaces/interfaces.ts

```diff
@@ -54,34 +30,143 @@ namespace interfaces {
+        DynamicValueProvider<TActivated> |
+        ConstructorValueProvider<TActivated> |
+        FactoryValueProvider<TActivated> |
+        ProviderValueProvider<TActivated> |
+        NotConfiguredValueProvider
+
+    export type BindingType = ValueProviderType<unknown>["type"];
+
+    export interface BindingTypeEnum {
+        ConstantValue: ConstantValueProvider<unknown>["type"];
+        Constructor: ConstructorValueProvider<unknown>["type"];
+        DynamicValue: DynamicValueProvider<unknown>["type"];
+        Factory: FactoryValueProvider<unknown>["type"];
+        Instance: InstanceValueProvider<unknown>["type"];
+        Provider: ProviderValueProvider<unknown>["type"];
+        NotConfigured: NotConfiguredValueProvider["type"];
+    }
+
+    export interface Scope<T>{
+        get(binding:Binding<T>,request:Request):Promise<T>|T|undefined
+        set(binding:interfaces.Binding<T>,request:Request,resolved:T|Promise<T>):T | Promise<T>
+    }
+
+    export interface SingletonScope<T> extends Scope<T>, Clonable<SingletonScope<T>>{
+        type:"Singleton",
+        resolved: T | Promise<T> | undefined;
+    }
+
+    export interface TransientScope<T> extends Scope<T>,Clonable<TransientScope<T>>{
+        type:"Transient"
+    }
+
+    export interface RequestResolveScope<T> extends Scope<T>, Clonable<RequestResolveScope<T>>{
+        type:"Request"
+    }
+
+    export interface RootRequestScope<T> extends Scope<T>,Clonable<RootRequestScope<T>>{
+        type:"RootRequest"
+    }
+
+    export interface CustomScope<T> extends Scope<T>,Clonable<CustomScope<T>>{
+        type:"Custom"
+    }
+
+    export interface NotConfiguredScope extends Scope<never>,Clonable<NotConfiguredScope>{
+        type:"NotConfigured",
+        get():never
+        set():never
+    }
+
+    export type BindingScopeScope<T> = SingletonScope<T> | TransientScope<T> | RequestResolveScope<T> | RootRequestScope<T>;
+    export type ResolveScope<T> =  BindingScopeScope<T> | CustomScope<T> | NotConfiguredScope;
+
+    export type BindingScope = BindingScopeScope<unknown>["type"];
+    export type ConfigurableBindingScope = ResolveScope<unknown>["type"];
+
+    export interface BindingScopeEnum {
+        Request: RequestResolveScope<any>["type"]
+        Singleton: SingletonScope<any>["type"];
+        Transient: TransientScope<any>["type"];
+        RootRequest: RootRequestScope<any>["type"];
+        Custom: CustomScope<any>["type"];
+        NotConfigured: NotConfiguredScope["type"];
+    }
+
+    export interface ConfigurableBindingScopeEnum extends BindingScopeEnum{
+        Custom: CustomScope<any>["type"];
+    }
+
     export interface Binding<TActivated> extends Clonable<Binding<TActivated>> {
         id: number;
         moduleId: ContainerModuleBase["id"];
-        activated: boolean;
         serviceIdentifier: ServiceIdentifier<TActivated>;
         constraint: ConstraintFunction;
-        dynamicValue: DynamicValue<TActivated> | null;
```

Comment:
dyamicValue, implementationType, factory, provider, cache ( in terms of constant value ) are now valueFrom on the ValueProvider